src/QsCompiler/Transformations/Monomorphization.cs

```diff
@@ -87,13 +128,97 @@ public override TypedExpression Transform(TypedExpression ex)
+
+            NonNullable<string> name;
+
+            if (concretions.ContainsKey(target))
+            {
+                name = concretions[target].FullName.Name;
+            }
+            else
+            {
+                // Create new name
+                string nameString = Guid.NewGuid().ToString() + "_" + globalCallable.Item.Name.Value;
+
+                // TODO: Check for identifier name collisions (not likely)
+                // - check in global namespace
+                // - check in concretion dict for current generic
+
+                name = NonNullable<string>.New(nameString);
+                
+                QsCallable newCallable = ReplaceTypeParamsSyntax.Apply(types, original)
+                    .WithFullName(fullname => new QsQualifiedName(globalCallable.Item.Namespace, name));
+
+                concretions.Add(target, newCallable);
+            }
+
+            return Identifier.NewGlobalCallable(new QsQualifiedName(globalCallable.Item.Namespace, name));
+        }
+    }
+
+    public class ResolveGenericsExpressionKind : ExpressionKindTransformation<ResolveGenericsExpression>
+    {
+        public ResolveGenericsExpressionKind(ResolveGenericsExpression expr) : base(expr) { }
+
+        public override QsExpressionKind<TypedExpression, Identifier, ResolvedType> onIdentifier(Identifier sym, QsNullable<ImmutableArray<ResolvedType>> tArgs)
+        {
+            return base.onIdentifier(sym, tArgs);
+        }
+    }
+
+    #endregion
+
+    #region ReplaceTypeParams
+
+    internal class ReplaceTypeParamsSyntax :
+        SyntaxTreeTransformation<MinorTransformations.ReplaceTypeParams>
+    {
+        public static QsCallable Apply(ImmutableDictionary<QsTypeParameter, ResolvedType> typeParams, QsCallable callable)
+        {
+            if (callable == null) throw new ArgumentNullException(nameof(callable));
+
+            var filter = new ReplaceTypeParamsSyntax(new MinorTransformations.ReplaceTypeParams(typeParams));
+
+            // Create a holding namespace to use transform on
+            var ns = new QsNamespace(NonNullable<string>.New(""), ImmutableArray.Create(QsNamespaceElement.NewQsCallable(callable)), Enumerable.Empty<ImmutableArray<string>>().ToLookup(x => NonNullable<string>.New("")) );
+            ns = filter.Transform(ns);
+
+            // Extract back out the transformed callable
+            return ((QsNamespaceElement.QsCallable)ns.Elements.First()).Item;
+        }
+
+        public ReplaceTypeParamsSyntax(MinorTransformations.ReplaceTypeParams scope) : base(scope) { }
+
+        public override ResolvedSignature onSignature(ResolvedSignature s)
         {
-            return "_" + String.Join("", typeDict
-                .ToList()
-                .OrderBy(kvp => kvp.Key.TypeName)
-                .Select(kvp => kvp.Key.TypeName.Value + "_" + kvp.Value.Resolution.ToString() + "_")
+            // Remove the type parameters from the signature
```

Comment:
Where is the logic that replace the now concretized type parameters in the argument and return type?

Reply:
The inherited ReplaceTypeParams transformation acts on all type resolution nodes, including the nodes found under the signiture's ArgumentType and ReturnType nodes.